hs-bindgen/src-internal/HsBindgen.hs

```diff
@@ -177,25 +233,20 @@ writeTests testDir = do
   Helpers
 -------------------------------------------------------------------------------}
 
-write :: String -> Maybe FilePath -> String -> ArtefactM ()
+write :: String -> Maybe FilePath -> String -> Artefact ()
 write _    Nothing     str = liftIO $ putStrLn str
-write what (Just path) str = do
-      tracer <- artefactTracer <$> ask
-      traceWith tracer $ RunArtefactWriteFile what path
-      liftIO $ do
-        createDirectoryIfMissing True $ takeDirectory path
-        writeFile path str
+write what (Just path) str = FileWrite what path (TextContent str)
 
 writeByCategory ::
      String
   -> FilePath
   -> BaseModuleName
   -> ByCategory String
-  -> ArtefactM ()
+  -> Artefact ()
 writeByCategory what hsOutputDir moduleBaseName =
-    mapM_ (uncurry writeCategory) . Map.toList . unByCategory
+    Foldable.foldl' (>>) (pure ()) . map (uncurry writeCategory) . Map.toList . unByCategory
```

Comment:
Is there a reason why you have removed `mapM_`? I think it is nicer than a fold.